btree/btree.c

```diff
@@ -2231,7 +2275,10 @@ static void generic_move(struct nd *src, struct nd *tgt,
 #ifndef __KERNEL__
 static bool m0_btree_overflow_is_possible(const struct nd *node)
```

Comment:
In case of variable key size, we need to consider max key, value size  and we can figured out in m0_btree_overflow_is_possible itself ,whether it is fixed format or variable format, For example, if it is fixed format we can get key size, val size from node, else we will consider max values for determine  overflow. If we take key size and value size as an parameter we need to write check every time if it is fixed format or not and then pass the key , value size accordingly.

Reply:
In fixed/variable format you know the ksize and vsize as provided to the put function. It is difficult to add a concept of max key/value-size for variable key/value since we never know what that value could be. So it is better to get the key size and value size as provided in the rec structure parameter to put function.

Reply:
This will work for leaf node, but for internal nodes, if there is splitting , key can be different from the key provided in rec structure, so in this case we need to consider maximum key

Reply:
Are we sure the key in the leaf node will be different than the key in internal node ? I thought the value would be different in both the node types.

Reply:
Yes. Lets take an example, once we reach leaf node, and we find that there is an overflow based on given ksize, vsize from record structure provided by user, once we split the leaf node into left and right leaf node, the key which needs to get inserted at internal node will be the 1st key from right leaf node, so here the key which needs to get inserted at the parent node can be different from key provided by user.

Reply:
ok in that case when you are doing a split at the level below you can find the max key from all the available keys in the child node since one of them would be the key that would be placed in the parent node.

Reply:
Yes we can do that, Currently , I haven't added any logic for variable key, value size nodes, as some api's may required and depending on implementation of variable key value format, logic may changed, For example, to find max key from node, we may need api which gives size of key, in fixed format , as there is no such things we do not want any such api, and this api is not implemented yet. So, this things needs to be discussed. I have changed m0_btree_overflow_is_possible routine and now it is taking key value size as a parameter

Reply:
I think that the code will become difficult since you need to go through every key and get its size. Let us see if there is another quick way to do this or for now you can fallback to your max KEY size. We will revisit this one again later, but you need to add a TODO comment and mention about getting the correct max key size at runtime.

Reply:
okay I will add comments at line . no. 2925 where we are initializing key value size for variable node format

Reply:
Added comments at line no. 2925